src/Unit-2/lesson4/README.md

```diff
@@ -0,0 +1,292 @@
+# Unit 2, Lesson 4 - Mapping(, filtering) and reducing 
+
+Hello again!
+
+In the [previous lesson](../lesson3/README.md), you learned how to create
+multi-map indexes and that was amazing. Right?
+
+In this lesson you will learn how to create `Map-Reduce indexes`. 
+
+## What is Map-Reduce?
+
+In essence, it is just a way to take a big task and divide it into discrete
+tasks that can be done in parallel. 
+
+A Map-Reduce process is composed of a `Map` function that projects data from
+documents into a common output (expected) and a `Reduce` function that performs
+a summary operation.
+
+I strongly recommend you to read this [blog post](https://ayende.com/blog/4435/map-reduce-a-visual-explanation). 
+Go ahead! I will be waiting for you.
+
+Still confused? Let's write some code and make it clearer.
+
+## Excercise: Your first map-reduce index
+
+I think the best way to learn about Map-Reduce indexes are writing some code.
+
+In this first excercise let's create perform a very simple task. Let's just
+count the number of products for each category. 
+
+Again, let's do it using the C# API.
+
+### Step 1: Create a new project and install the latest `RavenDB.Client` package
+
+Start Visual Studio and create a new `Console Application Project` named
+`MapReduceIndexes`. Then, in the `Package Manager Console`, issue the following 
+command: 
+
+```Install-Package RavenDB.Client```
+
+This will install the latest RavenDB.Client binaries, which you will need in order
+to compile your code.
+
+Then you will need to add the `using` name space at the top of the `Program.cs`:
+
+````csharp
+using Raven.Client.Document;
+````
+
+### Step 2: Write the model classes
+
+Again and it will never be enough,  you don't need to write "complete" model classes when you are only reading
+from the database. 
+
+````csharp
+class Category
+{
+    public string Name { get; }
+}
+
+class Product
+{
+    public string Category { get; }
+}
+````
+
+Yes! Everything we need is here. In the `Product` documents, the category
+is specified by the Id in the `Category` property.
+
+### Step 3: Write the `Map-Reduce index definition`
+
+One way to create a Map-Reduce index  definition is inheriting from `AbstractIndexCreationTask`.
+
+In the [previous lesson](../lesson3/README.md), you learned how to create multi-map indexes.
+It's importante you know that you could can combine the power of multi-map with map-reduce just
+providing multiple map functions. 
+
+````csharp
+public class Products_ByCategory : AbstractIndexCreationTask<Product, Products_ByCategory.Result>
+{
+    public class Result
+    {
+        public string Category { get; set; }
+        public int Count { get; set; }
+    }
+
+    public Products_ByCategory()
+    {
+        Map = products => 
+            from product in products
+            let categoryName = LoadDocument<Category>(product.Category).Name
```

Comment:
You haven't talked about `LoadDocument` yet, and it is a somewhat tricky feature.
Better to group by the category id, then include / transform this on the way out.